Add UploadFileCommand for Nextcloud

NobodysNightmare opened the pull request:
# Ticket
https://community.openproject.org/wp/68864

NobodysNightmare committed:
Add UploadFileCommand for Nextcloud

Files will be uploaded by OpenProject directly in this case and
not through the user's browser.

NobodysNightmare commented on modules/storages/app/common/storages/adapters/input/upload_file_contract.rb:
```diff
@@ -0,0 +1,44 @@
+          required(:file_name).filled(:string)
+
+          # TODO: how to define that we expect an IO-compatible thing?
+          required(:io)
```
@mereghost Do you have any opinion on how to solve this?

I guess I could [match on the interface](https://dry-rb.org/gems/dry-types/1.7/custom-types/) somehow, but honestly: I don't know which part of the IO interface `httpx` relies, I just know that it accepts an `IO`-like object as a body.

NobodysNightmare replied:
One thing we just noticed, that's probably making easy solutions unfeasible:

```
pry(main)> StringIO.new.is_a?(IO)
=> false
```

There are legitimate IOs that do not inherit from IO :duck:

mereghost replied:
I'm ok with it if we at least add a `.filled` here to avoid passing `nil` as `required` just checks if the key is there, not if there's a value there. This will be deprecated later on, so I'd prefer the solution below.

We can also define a new `AdapterType`, using a `AdapterTypes.Instance(IO)` to ensure an IO compatible object.

mereghost replied:
As for `StringIO` if we want to support we can define a sum type:

```ruby
UploadFileIO = AdatperTypes.Instance(IO) | AdapterTypes.Instance(StringIO)
```

NobodysNightmare replied:
The [docs say](https://dry-rb.org/gems/dry-types/1.7/custom-types/#code-types-instance-code):

> Types.Instance builds a type that checks if a value has the given class.

As I outlined above, this is not usable for our use case. At first I thought that I'd need to make up a hypothetical class that could implement the IO interface without inheriting from IO (which is perfectly valid, as Ruby uses duck typing). Luckily, I don't have to make that case up, because the only class I tested file upload with so far is `StringIO`, which implements the `IO` interface for any given string. It works fine, but it does not inherit from `IO`.

And I don't know (and I don't want to restrict) which part of the IO interface `httpx` is using to make this stuff work.

What I understand is that `required(:io).filled` would be valid DSL to expect a non-nil value to be passed to `:io`.

NobodysNightmare replied:
> As for StringIO if we want to support we can define a sum type

I read that only after my answer (thanks GitHub for not dynamically loading new responses): I think this does not solve the deeper problem, that this approach does not support duck typing as well. To give one specific example: I could now not pass a `instance_double(IO)` in an RSpec test anymore. The same way, if there's any gem out there that implements its own IO-compatible object, we'd not support it either, unless it makes sure to explicitly inherit from `IO`.

mereghost replied:
> What I understand is that required(:io).filled would be valid DSL to expect a non-nil value to be passed to :io.

yep, filled with no argument will be deprecated soon-ish, that's why I'd avoid it.

> instance_double(IO)

I'm fine making this sacrifice for some consistency, specially since there are ways around using doubles.

NobodysNightmare commented on modules/storages/app/common/storages/adapters/adapter_types.rb:
```diff
@@ -39,6 +39,9 @@ module AdapterTypes
       HTTPVerb = AdapterTypes::Nominal::Symbol.constrained(included_in: %i(post put))
+
+      # Type requirements for an IO passed to HTTPX
+      HttpxIO = (AdapterTypes.Interface(:read) | AdapterTypes.Interface(:each))
```
Rubocop complains about my non-usage of SCREAMING_SNAKE_CASE.

Though, I guess it would complain about the rest of the class as well. Should we selectively disable this cop here?

mereghost commented on modules/storages/app/common/storages/adapters/providers/nextcloud/commands/upload_file_command.rb:
```diff
@@ -0,0 +1,80 @@
+
+              case response
+              in { status: 200..299 }
+                Success(response)
```
🔴 How does response look like? Can we build a ResultData::StorageFile out of it? I'm asking because we will need some of its properties to build a file link.

NobodysNightmare replied:
Let me cite the VCR cassette:

```
  response:
    status:
      code: 204
      message: No Content
```

mereghost replied:
Fuck. Like we do for create folder, we might need to do a file_info request over the newly updated file and return its result.

shiroginne replied:
If I'm getting it right, we need:
```ruby
        origin_id: file_info.id,
        origin_name: file_info.name,
        origin_mime_type: file_info.mime_type,
```

to create a file link

NobodysNightmare replied:
The pushed state will already return a proper result. Don't be surprised: The mime type is always indicated as a folder, because of stupid copy&paste.

I still need to cleanup and deduplicate the handling of these responses.

NobodysNightmare commented on modules/storages/app/common/storages/adapters/input/upload_file_contract.rb:
```diff
@@ -0,0 +1,42 @@
+    module Input
+      class UploadFileContract < DryApplicationContract
+        params do
+          required(:file_path).filled(:string)
```
I think we recently discussed that we want to split the input into "name of the file" and "path where to put the file", because storage providers might have different validation rules.

Should I do this for the command already? Should the path then be a `AdapterTypes::Location`?

mereghost replied:
Yup. It isn't great but at least becomes standardized.

NobodysNightmare committed:
Unify storage_file methods implemented in different places

This was highly duplicated code with incomplete and inconsistent
implementations across the board. Every implementation had own attributes
to add, that another implementation didn't care filling out and some
aspects, such as the escaping of the location were not consistent at all.

NobodysNightmare commented on modules/storages/app/common/storages/adapters/providers/nextcloud/storage_file_transformer.rb:
```diff
@@ -0,0 +1,96 @@
+
+            # TODO: This is utterly inconsistent; create_folder expected to create an unescaped representation (e.g. "a folder"),
+            #       files_query expected to create an escaped representation (e.g. "a%20folder"), but both generated the SAME
+            #       kind of result. Who's right?
```
@Kharonus @mereghost I want to meet the two of you and talk about this.

You can probably tell me which implementation is right and which implementation is wrong. To me it feels like `FilesQuery` is right, because it very explicitly tests for those cases. On the other hand a lot of stuff starts to break, if I change towards that.

NobodysNightmare replied:
Result of talking to @mereghost directly: Unescaped is probably correct, because that's what we decided on for most places (also OneDrive and SharePoint).

NobodysNightmare committed:
Share specs for command signatures even more

Those looked the same across all queries and commands,
so we share on that level. We still afford to keep different
examples for commands and queries.

NobodysNightmare commented on modules/storages/app/common/storages/adapters/results/storage_file_contract.rb:
```diff
@@ -43,7 +43,7 @@ class StorageFileContract < DryApplicationContract
           optional(:created_at).filled(:time)
           optional(:last_modified_at).filled(:time)
-          optional(:created_by_name).filled(:string)
+          optional(:created_by_name).value(:string)
```
CC @mereghost : The previous implementation of Nextcloud's `files_info` command bypassed this contract, thus we already had `StorageFile`s with am empty value here.

Now that I unified the parsing of `StorageFile`, all of them get through this contract, so I made the contract more lax.

Alternatively we could insert a default value (e.g. "Unknown"). WDYT?

mereghost replied:
Having a default could be nice. =)

NobodysNightmare replied:
Default it is.

NobodysNightmare committed:
Rename testing environment variable

NEXTCLOUD_LOCAL_GROUP_USER_PASSWORD was not really telling what
kind of password we are setting. The new name makes it more clear
that the password is related to the AMPF feature.
